Refactor WaitGroup implementation in tasks.go

shrutiparabgoogle opened the pull request:
Fixes https://github.com/apigee/registry/issues/239

Create a struct for taskQueues
Get rid of the shared variables and use instance specific attributes

shrutiparabgoogle commented:
The build will fail because the taskQueue change is yet to be applied to the rest of the commands

seaneganx reviewed (commented):
Using a lightweight worker function might do just as well with the same amount of work required by the caller. Something like [this example](https://gobyexample.com/worker-pools) combined with [this example](https://gobyexample.com/waitgroups).

Usage would be similar to our current setup:
```
var wg sync.WaitGroup
tasks := make(chan core.Task, 1024)
for i := 0; i < 16; i++ {
	wg.Add(1)
	go core.Worker(ctx, wg, tasks)
}
defer wg.Wait()

// Generate tasks.
tasks <- &computeLintTask{ ... }

// Let workers finish.
close(tasks)
```

We can also consider using a helper to handle most of the setup work. Something like `func Pool(ctx context.Context, n int) (queue chan<- Task, wait func())` should make everything easy for the caller.

```
queue, wait := Pool(ctx, 16)
defer wait()

// Generate tasks.
queue <- &computeLintTask{ ... }
```

shrutiparabgoogle commented:
The idea was to abstract out the setup and underlying details for creating taskQueues. Hence, I ended up creating the struct. That said, I liked the idea of the helper func which you proposed `func Pool(ctx context.Context, n int) (queue chan<- Task, wait func())` it also doing the same job without the complexity of a struct. Will make changes accordingly.

shrutiparabgoogle commented:
Update the PR with following changes:
1. Refactored waitgroup implementation
2. Migrated resolve command to use taskQueues and call the sub-commands programmatically. (These changes are mainly reflected in `local-tasks.go`).
3. While importing all the sub-command packages in `local-tasks.go`, there was a cyclic dependency because `upload` command was depending on controller package for `ReadManifestProto`. As follows:
`registry -> resolve -> controller -> upload (in local-tasks.go) -> controller`
Hence, got rid of `manifest-reader.go` and moved the funcs to respective packages.

seaneganx commented on cmd/registry/controller/local-tasks.go:
```diff
@@ -26,16 +38,57 @@ type ExecCommandTask struct {
+	}
+
+	switch action[0] {
+	case "annotate":
+		return annotate.Command(ctx), nil
+	case "compute":
+		return compute.Command(ctx), nil
+	case "delete":
+		return delete.Command(ctx), nil
+	case "export":
+		return export.Command(ctx), nil
+	case "get":
+		return get.Command(ctx), nil
+	case "index":
+		return index.Command(ctx), nil
+	case "label":
+		return label.Command(ctx), nil
+	case "list":
+		return list.Command(ctx), nil
+	case "search":
+		return search.Command(ctx), nil
+	case "upload":
+		return upload.Command(ctx), nil
+	case "vocabulary":
+		return vocabulary.Command(ctx), nil
```
If we ever change these subcommand names, will there be a test that points out this connection is broken?

shrutiparabgoogle replied:
Added another bug to take care to the tests https://github.com/apigee/registry/issues/248

seaneganx reviewed (commented):
Would that circular dependency issue still occur if the manifest reader was in its own package? That might be the better approach instead of copying the code into each package that needs it.

shrutiparabgoogle commented on cmd/registry/core/tasks.go:
```diff
@@ -26,18 +26,29 @@ type Task interface {
 }
 
-var wg sync.WaitGroup
+func WorkerPool(ctx context.Context, n int) (chan Task, func()) {
```
Done.

shrutiparabgoogle commented on cmd/registry/core/tasks.go:
```diff
@@ -26,18 +26,29 @@ type Task interface {
+		wg.Add(1)
+		go worker(ctx, &wg, taskQueue)
+	}
+	return taskQueue, wg.Wait
```
Done.

shrutiparabgoogle commented on cmd/registry/controller/controller_test.go:
```diff
@@ -168,6 +171,26 @@ func createUpdateArtifact(
 
+func readManifestProto(
+	filename string,
+	t *testing.T) (*rpc.Manifest, error) {
```
Not relevant anymore.

shrutiparabgoogle commented:
While evaluating the dependency on `manifest_reader` and your suggestion about creating a package outside of controller, I realized the funcs in manifest_reader are very specific to what `resolve.go` wants to do and not very generic to be used anywhere else. Also, `controller_test.go` was using `ReadManifestProto`, which is not actually needed. I replaced that usage with `rpc.Manifest` structs. We were not testing the `ReadManifestProto()` anywhere, so this new approach makes the test error proof as well. Also, I think it improves the time taken to run controller tests, since it doesn't need to do any IO operations any more.

shrutiparabgoogle commented on cmd/registry/cmd/annotate/annotate.go:
```diff
@@ -45,12 +45,8 @@ func Command(ctx context.Context) *cobra.Command {
-				go core.Worker(ctx, taskQueue)
-			}
+			taskQueue, closeAndWait := core.WorkerPool(ctx, 64)
+			defer closeAndWait()
```
Done.
